frb/analysis/dmhalos_photoz.py

```diff
@@ -0,0 +1,615 @@
+        p = Pool(n_cores)
+        func = lambda x: halomass_from_stellarmass(x, z = z, randomize=True)
+        log_mhalo_array = p.map(func, log_mstar)
+        
+        return log_mhalo_array
+
+
+    # Loop over log_mstar:
+    for idx, log_mstar in enumerate(log_mstar_array):
+        temp_log_mstar = np.full(n_halo, log_mstar)
+        log_mhalo_array[idx] = mhalo_factory(temp_log_mstar, z = z, n_cores = 10)
+    
+    # Store this in a fits file
+    prihdu  = fits.PrimaryHDU()
+    starhdu = fits.ImageHDU(log_mstar_array, name = "MSTAR")
+    halohdu = fits.ImageHDU(log_mhalo_array, name = "MHALO")
+    hdulist = fits.HDUList([prihdu, starhdu, halohdu])
+    hdulist.writeto(fits_out, overwrite = True)
+
+    return
+
+def mhalo_lookup_tables(z_grid:list, datafolder:str=DEFAULT_DATA_FOLDER):
+    """
+    For each z in z_grid, produces a fits file containing m_halo values
+    corresponding to a fixed grid of m_star values. The values are produced
+    by sampling the Moster+13 SHMR relation. The fits files can then be
+    used to produce interpolation functions of the moments of the m_halo
+    distribution (e.g. mean, std.dev) as a function of redshift and log_mstar.
+    Args:
+        z_grid (list or np.ndarray): List of redshift values to be sampled.
+        datafolder (str, optional): Path to the directory where the results will be stored.
+    """
+
+    # Just loop over z_grid and produce the fits files.
+    for z in z_grid:
+        realization_file = os.path.join(datafolder, "mhalo_realization_z_{:0.2f}.fits".format(z))
+        _mhalo_lookup_table(z, realization_file)
+
+    return
+
+def _mhalo_realizations(log_mstar:float, log_mstar_err:float, z:float,
+                        mean_interp:interp2d, stddev_interp:interp2d,
+                        n_mstar:int=100, n_norm:int=10)->np.ndarray:
+    """
+    Using the lookup tables generated (see function mhalo_lookup_tables), produce
+    realiztions of mhalo. This takes into account both the stellar mass uncertainty
+    and the uncertainty in the SMHR relation from Moster+13.
+    Args:
+        log_mstar (float): log stellar mass in M_sun.
+        log_mstar_err (float): log error in log_mstar
+        z (float): redshift
+        mean_interp (interp2d): <log_mhalo(log_mstar, z)> (based on SHMR)
+        stddev_interp (interp2d): std.dev. log_mhalo(log_mstar, z) (based on SHMR)
+        n_mstrar (int, optional): Number of m_star samples to be produced.
+        n_norm (int, optional): Number of m_halo samples for each m_star sample.
+    Returns:
+        mhalo_reals (np.ndarray): log_mhalo realizations.
+    """
+
+    # First produce realizations of mstar from a normal distribution.
+    mstar_reals = np.random.normal(log_mstar, log_mstar_err, n_mstar)
+
+    # Then get mean values of halo masses for each stellar mass.
+    mean_mhalo_reals = mean_interp(mstar_reals, z)
+    mean_mhalo_reals[mean_mhalo_reals>12.8] = 12.8 # Set a cutoff for the mean halo mass
```

Comment:
Just so you know, there is a `numpy.maximum` method that does this.

Reply:
Thanks! Replaced with the `np.minimum`!